src/shared/api/schemas.ts

```diff
@@ -0,0 +1,252 @@
+  if (accountParts.length !== 2) {
+    ctx.addIssue({
+      code: z.ZodIssueCode.invalid_literal,
+      message: 'Invalid accountId format',
+      expected: 'signatory@account_domain format',
+      received: val,
+    });
+  }
+
+  const [account, account_domain] = accountParts;
+
+  return new AccountId(account, account_domain as DomainId);
+});
+
+export class AssetDefinitionId {
+  public readonly name: string;
+  public readonly domain: DomainId;
+
+  public constructor(name: string, domain: DomainId) {
+    this.name = name;
+    this.domain = domain;
+  }
+
+  public toString() {
+    return this.name + '#' + this.domain;
+  }
+}
+
+export const AssetDefinitionIdSchema = z.string().transform((val, ctx) => {
+  const assetDefinitionParts = val.split('#');
+
+  if (assetDefinitionParts.length !== 2) {
+    ctx.addIssue({
+      code: z.ZodIssueCode.invalid_literal,
+      message: 'Invalid assetDefinitionId format',
+      expected: 'asset#asset_domain format',
+      received: val,
+    });
+  }
+
+  const [asset, asset_domain] = assetDefinitionParts;
+
+  return new AssetDefinitionId(asset, asset_domain as DomainId);
+});
+
+export class AssetId {
+  public readonly account: AccountId;
+  public readonly definition: AssetDefinitionId;
+
+  public constructor(account: AccountId, definition: AssetDefinitionId) {
+    this.account = new AccountId(account.signatory, account.domain);
+    this.definition = new AssetDefinitionId(definition.name, definition.domain ? definition.domain : account.domain);
+  }
+
+  public toString() {
+    if (this.account.domain === this.definition.domain) {
+      return this.definition.name + '##' + this.account.toString();
+    }
+
+    return this.definition.toString() + '#' + this.account.toString();
+  }
+}
+
+export const AssetIdSchema = z.string().transform((val, ctx) => {
+  const assetIdParts = val.split('#');
+  if (assetIdParts.length !== 3) {
+    ctx.addIssue({
+      code: z.ZodIssueCode.invalid_literal,
+      message: 'Invalid assetId format',
+      expected: 'asset#asset_domain#signatory@account_domain or asset##signatory@domain format',
+      received: val,
+    });
+  }
+
+  const [asset, asset_domain, account] = assetIdParts;
+
+  const accountParts = account.split('@');
+  if (accountParts.length !== 2) {
+    ctx.addIssue({
+      code: z.ZodIssueCode.invalid_literal,
+      message: 'Invalid account format',
+      expected: 'signatory@account_domain format',
+      received: account,
+    });
+  }
+
+  const [signatory, account_domain] = accountParts;
+
+  return new AssetId(
+    new AccountId(signatory, account_domain as DomainId),
+    new AssetDefinitionId(asset, asset_domain as DomainId)
+  );
+});
+
+export const Account = z.object({
+  id: AccountIdSchema,
+  metadata: Metadata,
+  owned_assets: z.number(),
+  owned_domains: z.number(),
+});
+
+export type Account = z.infer<typeof Account>;
+
+export interface AssetSearchParams extends PaginationParams {
+  owned_by?: string
```

Comment:
```suggestion
  owned_by?: AccountId
```

Stronger types!

Tip: if you also add `toJSON()` method to `AccountId` class with `toString()` inside, it will automatically stringify on serialisation without manual intervention.

Reply:
Thanks, thats useful. By the way, I haven't used until yet approach to pass an object in the url so it's automatically uses toString method and converts our classes into normal strings.